resources/http.go

```diff
@@ -0,0 +1,258 @@
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+)
+
+func init() {
+	RegisterResource("http", func() Res { return &HTTPRes{} })
+}
+
+// HTTPRes is an http resource for retrieving data over http
+type HTTPRes struct {
+	BaseRes        `yaml:",inline"`
+	URL            string            `yaml:"url"`
+	Timeout        int               `yaml:"timeout"` // the http timeout in seconds
+	Method         string            `yaml:"method"`
+	Header         map[string]string `yaml:"header"`
+	Body           string            `yaml:"body"`
+	responseBody   *string
+	responseCode   int
+	responseHeader http.Header
+	client         *http.Client
+}
+
+type httpResponse struct {
+	code   int
+	body   string
+	err    error
+	header http.Header
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *HTTPRes) Default() Res {
+	return &HTTPRes{
+		BaseRes: BaseRes{
+			MetaParams: DefaultMetaParams, // force a default
+		},
+		Method: http.MethodGet,
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *HTTPRes) Validate() error {
+	if obj.URL == "" { // this is the only thing that is really required
+		return fmt.Errorf("url can't be empty")
+	}
+
+	return obj.BaseRes.Validate()
+}
+
+// Init runs some startup code for this resource.
+func (obj *HTTPRes) Init() error {
+	obj.client = &http.Client{
+		Timeout: time.Duration(obj.Timeout) * time.Second,
+	}
+
+	obj.SetKind("http")
+	return obj.BaseRes.Init() // call base init, b/c we're overriding
+}
+
+// Watch is the primary listener for this resource and it outputs events.
+func (obj *HTTPRes) Watch() error {
+
+	if err := obj.Running(); err != nil {
+		return err // bubble up a NACK...
+	}
+
+	var send = false // send event?
+	var exit *error
+
+	wg := &sync.WaitGroup{}
+
+	httpResponseChan := make(chan *httpResponse)
+	closeChan := make(chan struct{})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			body := bytes.NewBufferString(obj.Body)
+			req, err := http.NewRequest(obj.Method, obj.URL, body)
+			if err != nil {
+				resp := httpResponse{0, "", err, nil}
```

Comment:
I think you can just do:
```
resp := httpResponse{err: err}
```
instead. Also it's preferable to do:

```
resp := &httpResponse{ ... }
```

instead of taking the pointer of resp. This way the object is created in it's correct pointer state at the source :)

This all applies in a few cases in the code...